src/dbus_proxy/_proxy.lua

```diff
@@ -235,6 +291,39 @@ local function generate_method(interface_name, method)
 
 end
 
+--- Generate an async method.
+-- @param[type=string] interface_name the interface name
+-- @param[type=lgi.Gio.DBusMethodInfo] method the [method info object](https://developer.gnome.org/gio/stable/gio-D-Bus-Introspection-Data.html#GDBusMethodInfo-struct)
+-- @return a function that wraps @{call_async} so the method can be called
+-- by passing the arguments as simple lua types.
+-- @see call
+local function generate_async_method(interface_name, method)
+  local args = {}
+  for _, arg in ipairs(method.in_args) do
+    args[#args + 1] = {type = arg.signature}
+  end
+
+  return function (proxy, user_callback, user_data, ...)
+    assert(#{...} == #args,
+           string.format(
+             "Expected %d parameters but got %d",
+             #args, #{...}))
+
+    for idx, val in ipairs({...}) do
+      args[idx].value = val
+    end
```

Comment:
This code seems to be the same as the implementation in `generate_method`, can you extract it to a local function that can be reused by both please?

Something like the below (note that I have not tested that it actually works!)

``` lua
local function generate_method(interface_name, method_info, is_async)
  local args = {}
  for _, arg in ipairs(method.in_args) do
    args[#args + 1] = {type = arg.signature}
  end

  return function (proxy, ...)
    assert(#{...} == #args,
           string.format(
             "Expected %d parameters but got %d",
             #args, #{...}))

    for idx, val in ipairs({...}) do
      args[idx].value = val
    end

    if is_async then

        return call_async(
      proxy,
      interface_name,
      method.name,
      user_callback,
      user_data,
      args)

    end

    return call(
      proxy,
      interface_name,
      method.name,
      args)
  end
end

local function generate_sync_method(...)
    return generate_method(..., false)
end

local function generate_async_method(...)
    return generate_method(..., true)
end
```

Reply:
Sure, thanks for the example!

Reply:
I see we have the `build_param` function as well which is called afterwards with the arguments table which in turn returns the GVariant, maybe `build_param` and `build_args` could be combined?

Reply:
I'd prefer to keep `build_params` as a separate function because its code is not trivial, but I'm happy to factor out `build_args` and use `build_params` inside it, then have `generate_...` and `generate_async_...`